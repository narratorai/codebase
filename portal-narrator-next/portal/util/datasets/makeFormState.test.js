import makeFormState from './makeFormState'

import datasetFullResponse from '../../../test/fixtures/datasetFullResponse.json'

const tables = [
  {
    customer_table: 'v_customers',
    customer_label: 'People',
    activity_stream: 'mv_activity_stream',
    identifier: 'Email',
  },
]

const liveActivities = [
  {
    slug: 'survey_email_captured',
    enrichment: {
      table: 'mv_enrich_uranium',
    },
  },
]

describe('#makeFormState', () => {
  it('creates Form State from queryDefinition', () => {
    expect(makeFormState({ queryDefinition: datasetFullResponse.query_definition })).toMatchSnapshot()
  })

  it('updates the enrichment in the query definition based on the live activities', () => {
    expect(makeFormState({ queryDefinition: datasetFullResponse.query_definition, liveActivities })).toMatchSnapshot()
  })

  it('creates default form from empty queryDefinition', () => {
    expect(makeFormState({ selectedActivityStream: 'mv_activity_stream', configTables: tables })).toMatchSnapshot()
  })

  it('throws on an empty queryDefinition without a selected activity stream', () => {
    expect(() => makeFormState({ configTables: tables })).toThrowErrorMatchingSnapshot()
  })
})
