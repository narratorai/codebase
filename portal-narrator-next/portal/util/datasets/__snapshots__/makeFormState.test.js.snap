// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#makeFormState creates Form State from queryDefinition 1`] = `
{
  "_attribute_activities": [
    {
      "_columns": [
        {
          "filters": [],
          "id": "photos_received_at",
          "label": "photos_received_at",
          "name": "ts",
          "output": true,
          "source_details": {
            "activity_kind": "append",
            "activity_slug": "photos_recieved",
          },
          "source_kind": "activity",
          "type": "timestamp",
        },
      ],
      "config": {
        "activity_stream": "mv_activity_stream",
        "customer_table": "v_customers",
        "enrichment_column": null,
        "enrichment_table": null,
        "has_source": true,
      },
      "did": true,
      "filters": [],
      "kind": "append",
      "occurence_after_time": false,
      "occurrence": "first",
      "occurrence_value": 1,
      "relationships": [],
      "slug": "photos_recieved",
    },
  ],
  "_computed_columns": [
    {
      "filters": [],
      "id": "session_attribution",
      "label": "session_attribution",
      "name": "",
      "output": true,
      "source_details": {
        "column": "session_source",
        "delimiter": " - ",
        "kind": "string_concatenate_multi_column",
        "second_column": "session_kind",
      },
      "source_kind": "computed",
      "type": "string",
    },
    {
      "filters": [],
      "id": "sa_row_number",
      "label": "sa_row_number",
      "name": "sa_row_number",
      "output": true,
      "source_details": {
        "full_order": [
          "first_session_at",
        ],
        "group": [
          "session_attribution",
        ],
        "kind": "row_number",
      },
      "source_kind": "computed",
      "type": "number",
    },
    {
      "filters": [],
      "id": "sa_count",
      "label": "sa_count",
      "name": "sa_count",
      "output": true,
      "source_details": {
        "group": [
          "session_attribution",
        ],
        "kind": "count",
      },
      "source_kind": "computed",
      "type": "number",
    },
    {
      "filters": [],
      "id": "sa_percent_of_total",
      "label": "sa_percent_of_total",
      "name": "sa_percent_of_total",
      "output": true,
      "source_details": {
        "column": "sa_count",
        "group": [
          "session_attribution",
        ],
        "kind": "percent_of_total",
      },
      "source_kind": "computed",
      "type": "number",
    },
    {
      "filters": [],
      "id": "categorized_source",
      "label": "categorized_source",
      "name": "",
      "output": true,
      "source_details": {
        "cases": [
          {
            "column": "session_kind",
            "filter": {
              "operator": "=",
              "value": "Referral",
            },
            "value": "NOT Direct-Organic-Search",
            "value_kind": "string",
          },
          {
            "column": "session_source",
            "filter": {
              "operator": "=",
              "value": "Organic",
            },
            "value": "Direct-Organic-Search",
            "value_kind": "string",
          },
          {
            "column": "session_kind",
            "filter": {
              "operator": "is_not_null",
              "value": null,
            },
            "value": "is set test",
            "value_kind": "string",
          },
          {
            "column": "session_kind",
            "filter": {
              "operator": "in",
              "value": [
                "Direct",
                "Search",
              ],
            },
            "value": "****Direct-Organic-Search",
            "value_kind": "string",
          },
        ],
        "kind": "ifttt",
        "value": "NOT Direct-Organic-Search",
        "value_kind": "string",
      },
      "source_kind": "computed",
      "type": "string",
    },
    {
      "filters": [],
      "id": "decimated_session",
      "label": "decimated_session",
      "name": "",
      "output": true,
      "source_details": {
        "column": "first_session_at",
        "kind": "time_truncate",
        "resolution": 5,
        "segmentation": "minute",
      },
      "source_kind": "computed",
      "type": "timestamp",
    },
  ],
  "_conversion_activities": [
    {
      "_columns": [
        {
          "filters": [],
          "id": "survey_wizard_completed_at",
          "label": "survey_wizard_completed_at",
          "name": "ts",
          "output": true,
          "source_details": {
            "activity_kind": "conversion",
            "activity_slug": "survey_wizard_completed",
          },
          "source_kind": "activity",
          "type": "timestamp",
        },
        {
          "filters": [],
          "id": "converted_to_survey_wizard_completed_UUID",
          "label": "converted_to_survey_wizard_completed",
          "name": "",
          "output": true,
          "source_details": {
            "activity_kind": "conversion",
            "activity_slug": "survey_wizard_completed",
            "cases": [
              {
                "column": "survey_wizard_completed_at",
                "filter": {
                  "operator": "is_not_null",
                  "value": null,
                },
                "value": 1,
                "value_kind": "number",
              },
            ],
            "kind": "ifttt",
            "value": 0,
            "value_kind": "number",
          },
          "source_kind": "computed",
          "type": "number",
        },
      ],
      "config": {
        "activity_stream": "mv_activity_stream",
        "customer_table": "v_customers",
        "enrichment_column": null,
        "enrichment_table": null,
        "has_source": true,
      },
      "did": true,
      "filters": [],
      "id": "survey_wizard_completed_UUID",
      "kind": "conversion",
      "occurence_after_time": false,
      "occurrence": "first",
      "occurrence_value": 1,
      "relationships": [],
      "slug": "survey_wizard_completed",
    },
    {
      "_columns": [
        {
          "filters": [],
          "id": "ordered_impression_kit_at",
          "label": "ordered_impression_kit_at",
          "name": "ts",
          "output": true,
          "source_details": {
            "activity_kind": "conversion",
            "activity_slug": "ordered_impression_kit",
          },
          "source_kind": "activity",
          "type": "timestamp",
        },
        {
          "filters": [],
          "id": "converted_to_ordered_impression_kit_UUID",
          "label": "converted_to_ordered_impression_kit",
          "name": "",
          "output": true,
          "source_details": {
            "activity_kind": "conversion",
            "activity_slug": "ordered_impression_kit",
            "cases": [
              {
                "column": "ordered_impression_kit_at",
                "filter": {
                  "operator": "is_not_null",
                  "value": null,
                },
                "value": 1,
                "value_kind": "number",
              },
            ],
            "kind": "ifttt",
            "value": 0,
            "value_kind": "number",
          },
          "source_kind": "computed",
          "type": "number",
        },
        {
          "filters": [],
          "id": "days_from_session_to_ordered_impression_kit",
          "label": "days_from_session_to_ordered_impression_kit",
          "name": "",
          "output": true,
          "source_details": {
            "activity_kind": "conversion",
            "activity_slug": "ordered_impression_kit",
            "column": "first_session_at",
            "kind": "time_between",
            "second_column": "ordered_impression_kit_at",
            "segmentation": "days",
          },
          "source_kind": "computed",
          "type": "number",
        },
      ],
      "config": {
        "activity_stream": "mv_activity_stream",
        "customer_table": "v_customers",
        "enrichment_column": null,
        "enrichment_table": null,
        "has_source": true,
      },
      "did": true,
      "filters": [],
      "id": "ordered_impression_kit_UUID",
      "kind": "conversion",
      "occurence_after_time": false,
      "occurrence": "first",
      "occurrence_value": 1,
      "relationships": [],
      "slug": "ordered_impression_kit",
    },
  ],
  "_customer_table_columns": [
    {
      "filters": [],
      "id": "customer_name",
      "label": "name",
      "name": "name",
      "output": true,
      "source_details": {
        "table": "v_customers",
      },
      "source_kind": "customer",
      "type": "string",
    },
  ],
  "_limiting_activities": [
    {
      "_columns": [
        {
          "filters": [],
          "id": "email",
          "label": "email",
          "name": "customer",
          "output": true,
          "source_details": {
            "activity_kind": "limiting",
            "activity_slug": "session",
          },
          "source_kind": "activity",
          "type": "string",
        },
        {
          "filters": [],
          "id": "segment_id",
          "label": "segment_id",
          "name": "source_id",
          "output": true,
          "source_details": {
            "activity_id": "session_UUID",
            "activity_kind": "limiting",
          },
          "source_kind": "activity",
          "type": "string",
        },
        {
          "filters": [
            {
              "operator": ">",
              "value": "2018-10-01",
            },
          ],
          "id": "first_session_at",
          "label": "first_session_at",
          "name": "ts",
          "output": true,
          "source_details": {
            "activity_kind": "limiting",
            "activity_slug": "session",
          },
          "source_kind": "activity",
          "type": "timestamp",
        },
        {
          "filters": [],
          "id": "id_device",
          "label": "device",
          "name": "device",
          "output": true,
          "source_details": {
            "activity_kind": "limiting",
            "activity_slug": "session",
          },
          "source_kind": "enrichment",
          "type": "string",
        },
        {
          "filters": [],
          "id": "session_source",
          "label": "session_source",
          "name": "session_source",
          "output": true,
          "source_details": {
            "activity_kind": "limiting",
            "activity_slug": "session",
          },
          "source_kind": "enrichment",
          "type": "string",
        },
        {
          "filters": [],
          "id": "session_kind",
          "label": "session_kind",
          "name": "session_kind",
          "output": true,
          "source_details": {
            "activity_kind": "limiting",
            "activity_slug": "session",
          },
          "source_kind": "enrichment",
          "type": "string",
        },
      ],
      "config": {
        "activity_stream": "mv_activity_stream",
        "customer_table": "v_customers",
        "enrichment_column": "enriched_activity_id",
        "enrichment_table": "mv_enriched_pages",
        "has_source": true,
      },
      "did": true,
      "filters": [],
      "id": "session_UUID",
      "kind": "limiting",
      "occurence_after_time": false,
      "occurrence": "first",
      "occurrence_value": 1,
      "relationships": [],
      "slug": "session",
    },
    {
      "_columns": [
        {
          "filters": [],
          "id": "survey_email_captured_at",
          "label": "survey_email_captured_at",
          "name": "ts",
          "output": true,
          "source_details": {
            "activity_kind": "limiting",
            "activity_slug": "survey_email_captured",
          },
          "source_kind": "activity",
          "type": "timestamp",
        },
      ],
      "config": {
        "activity_stream": "mv_activity_stream",
        "customer_table": "v_customers",
        "enrichment_column": null,
        "enrichment_table": null,
        "has_source": true,
      },
      "did": true,
      "filters": [],
      "kind": "limiting",
      "occurence_after_time": false,
      "occurrence": "first",
      "occurrence_value": 1,
      "relationships": [],
      "slug": "survey_email_captured",
    },
  ],
  "_staged_activity": {
    "_columns": [],
    "_fieldIndex": 0,
    "_isEdit": false,
    "config": {},
    "did": true,
    "filters": [],
    "kind": "limiting",
    "occurence_after_time": false,
    "occurrence": "all",
    "occurrence_value": 1,
    "relationships": [],
    "slug": null,
  },
  "_staged_computed_column": {
    "_auto_generated_by": null,
    "filters": [],
    "id": null,
    "label": "",
    "name": "",
    "output": true,
    "source_details": {},
    "source_kind": "computed",
    "type": "string",
  },
  "_staged_filters": [
    {
      "kind": "value",
      "operator": "",
      "or_null": false,
      "value": "",
    },
  ],
  "_staged_group_by": {
    "_column_ids": [],
    "columns": [],
    "computed_columns": [],
    "metrics": [
      {
        "_auto_generated_by": null,
        "_pre_pivot_column_id": undefined,
        "_pre_pivot_column_label": undefined,
        "agg_function": "COUNT_ALL",
        "column_id": null,
        "filters": [],
        "id": "metrics_total_events",
        "label": "Count Records",
        "output": true,
        "pivot": [],
        "type": "number",
      },
    ],
    "name": "",
    "order": [],
    "parent_filters": [],
    "pivot": [],
    "slug": null,
  },
  "_staged_group_by_metric": {
    "_auto_generated_by": null,
    "_pre_pivot_column_id": undefined,
    "_pre_pivot_column_label": undefined,
    "agg_function": "COUNT_ALL",
    "column_id": null,
    "filters": [],
    "id": "metrics_total_events",
    "label": "Count Records",
    "output": true,
    "pivot": [],
    "type": "number",
  },
  "_staged_order_by": [],
  "activity_stream": "mv_activity_stream",
  "all_groups": [
    {
      "column_ids": [
        "device",
      ],
      "computed_columns": [],
      "metrics": [
        {
          "agg_function": "COUNT",
          "column_id": "*",
          "filters": [],
          "id": "metrics_total_events",
          "label": "total_events",
          "pivot": [],
        },
        {
          "agg_function": "SUM",
          "column_id": "converted_to_survey_wizard_completed",
          "filters": [],
          "id": "metrics_total_survey_wizard_completed",
          "label": "total_survey_wizard_completed",
          "pivot": [],
        },
      ],
      "name": "Group by Device",
      "order": [
        {
          "column_id": "metrics_total_events",
          "order_direction": "asc",
        },
      ],
      "pivot": [],
      "slug": "group_by_device",
    },
  ],
  "order": [
    {
      "column_id": "decimated_session",
      "order_direction": "desc",
    },
  ],
}
`;

exports[`#makeFormState creates default form from empty queryDefinition 1`] = `
{
  "_attribute_activities": [],
  "_computed_columns": [],
  "_conversion_activities": [],
  "_customer_table_columns": [],
  "_limiting_activities": [],
  "_staged_activity": {
    "_columns": [],
    "_fieldIndex": 0,
    "_isEdit": false,
    "config": {},
    "did": true,
    "filters": [],
    "kind": "limiting",
    "occurence_after_time": false,
    "occurrence": "all",
    "occurrence_value": 1,
    "relationships": [],
    "slug": null,
  },
  "_staged_computed_column": {
    "_auto_generated_by": null,
    "filters": [],
    "id": null,
    "label": "",
    "name": "",
    "output": true,
    "source_details": {},
    "source_kind": "computed",
    "type": "string",
  },
  "_staged_filters": [
    {
      "kind": "value",
      "operator": "",
      "or_null": false,
      "value": "",
    },
  ],
  "_staged_group_by": {
    "_column_ids": [],
    "columns": [],
    "computed_columns": [],
    "metrics": [
      {
        "_auto_generated_by": null,
        "_pre_pivot_column_id": undefined,
        "_pre_pivot_column_label": undefined,
        "agg_function": "COUNT_ALL",
        "column_id": null,
        "filters": [],
        "id": "metrics_total_events",
        "label": "Count Records",
        "output": true,
        "pivot": [],
        "type": "number",
      },
    ],
    "name": "",
    "order": [],
    "parent_filters": [],
    "pivot": [],
    "slug": null,
  },
  "_staged_group_by_metric": {
    "_auto_generated_by": null,
    "_pre_pivot_column_id": undefined,
    "_pre_pivot_column_label": undefined,
    "agg_function": "COUNT_ALL",
    "column_id": null,
    "filters": [],
    "id": "metrics_total_events",
    "label": "Count Records",
    "output": true,
    "pivot": [],
    "type": "number",
  },
  "_staged_order_by": [],
  "activity_stream": "mv_activity_stream",
  "all_groups": [],
  "order": [],
}
`;

exports[`#makeFormState throws on an empty queryDefinition without a selected activity stream 1`] = `"Unknown stream: Unable to make dataset form state"`;

exports[`#makeFormState updates the enrichment in the query definition based on the live activities 1`] = `
{
  "_attribute_activities": [
    {
      "_columns": [
        {
          "filters": [],
          "id": "photos_received_at",
          "label": "photos_received_at",
          "name": "ts",
          "output": true,
          "source_details": {
            "activity_kind": "append",
            "activity_slug": "photos_recieved",
          },
          "source_kind": "activity",
          "type": "timestamp",
        },
      ],
      "config": {
        "activity_stream": "mv_activity_stream",
        "customer_table": "v_customers",
        "enrichment_column": null,
        "enrichment_table": null,
        "has_source": true,
      },
      "did": true,
      "filters": [],
      "kind": "append",
      "occurence_after_time": false,
      "occurrence": "first",
      "occurrence_value": 1,
      "relationships": [],
      "slug": "photos_recieved",
    },
  ],
  "_computed_columns": [
    {
      "filters": [],
      "id": "session_attribution",
      "label": "session_attribution",
      "name": "",
      "output": true,
      "source_details": {
        "column": "session_source",
        "delimiter": " - ",
        "kind": "string_concatenate_multi_column",
        "second_column": "session_kind",
      },
      "source_kind": "computed",
      "type": "string",
    },
    {
      "filters": [],
      "id": "sa_row_number",
      "label": "sa_row_number",
      "name": "sa_row_number",
      "output": true,
      "source_details": {
        "full_order": [
          "first_session_at",
        ],
        "group": [
          "session_attribution",
        ],
        "kind": "row_number",
      },
      "source_kind": "computed",
      "type": "number",
    },
    {
      "filters": [],
      "id": "sa_count",
      "label": "sa_count",
      "name": "sa_count",
      "output": true,
      "source_details": {
        "group": [
          "session_attribution",
        ],
        "kind": "count",
      },
      "source_kind": "computed",
      "type": "number",
    },
    {
      "filters": [],
      "id": "sa_percent_of_total",
      "label": "sa_percent_of_total",
      "name": "sa_percent_of_total",
      "output": true,
      "source_details": {
        "column": "sa_count",
        "group": [
          "session_attribution",
        ],
        "kind": "percent_of_total",
      },
      "source_kind": "computed",
      "type": "number",
    },
    {
      "filters": [],
      "id": "categorized_source",
      "label": "categorized_source",
      "name": "",
      "output": true,
      "source_details": {
        "cases": [
          {
            "column": "session_kind",
            "filter": {
              "operator": "=",
              "value": "Referral",
            },
            "value": "NOT Direct-Organic-Search",
            "value_kind": "string",
          },
          {
            "column": "session_source",
            "filter": {
              "operator": "=",
              "value": "Organic",
            },
            "value": "Direct-Organic-Search",
            "value_kind": "string",
          },
          {
            "column": "session_kind",
            "filter": {
              "operator": "is_not_null",
              "value": null,
            },
            "value": "is set test",
            "value_kind": "string",
          },
          {
            "column": "session_kind",
            "filter": {
              "operator": "in",
              "value": [
                "Direct",
                "Search",
              ],
            },
            "value": "****Direct-Organic-Search",
            "value_kind": "string",
          },
        ],
        "kind": "ifttt",
        "value": "NOT Direct-Organic-Search",
        "value_kind": "string",
      },
      "source_kind": "computed",
      "type": "string",
    },
    {
      "filters": [],
      "id": "decimated_session",
      "label": "decimated_session",
      "name": "",
      "output": true,
      "source_details": {
        "column": "first_session_at",
        "kind": "time_truncate",
        "resolution": 5,
        "segmentation": "minute",
      },
      "source_kind": "computed",
      "type": "timestamp",
    },
  ],
  "_conversion_activities": [
    {
      "_columns": [
        {
          "filters": [],
          "id": "survey_wizard_completed_at",
          "label": "survey_wizard_completed_at",
          "name": "ts",
          "output": true,
          "source_details": {
            "activity_kind": "conversion",
            "activity_slug": "survey_wizard_completed",
          },
          "source_kind": "activity",
          "type": "timestamp",
        },
        {
          "filters": [],
          "id": "converted_to_survey_wizard_completed_UUID",
          "label": "converted_to_survey_wizard_completed",
          "name": "",
          "output": true,
          "source_details": {
            "activity_kind": "conversion",
            "activity_slug": "survey_wizard_completed",
            "cases": [
              {
                "column": "survey_wizard_completed_at",
                "filter": {
                  "operator": "is_not_null",
                  "value": null,
                },
                "value": 1,
                "value_kind": "number",
              },
            ],
            "kind": "ifttt",
            "value": 0,
            "value_kind": "number",
          },
          "source_kind": "computed",
          "type": "number",
        },
      ],
      "config": {
        "activity_stream": "mv_activity_stream",
        "customer_table": "v_customers",
        "enrichment_column": null,
        "enrichment_table": null,
        "has_source": true,
      },
      "did": true,
      "filters": [],
      "id": "survey_wizard_completed_UUID",
      "kind": "conversion",
      "occurence_after_time": false,
      "occurrence": "first",
      "occurrence_value": 1,
      "relationships": [],
      "slug": "survey_wizard_completed",
    },
    {
      "_columns": [
        {
          "filters": [],
          "id": "ordered_impression_kit_at",
          "label": "ordered_impression_kit_at",
          "name": "ts",
          "output": true,
          "source_details": {
            "activity_kind": "conversion",
            "activity_slug": "ordered_impression_kit",
          },
          "source_kind": "activity",
          "type": "timestamp",
        },
        {
          "filters": [],
          "id": "converted_to_ordered_impression_kit_UUID",
          "label": "converted_to_ordered_impression_kit",
          "name": "",
          "output": true,
          "source_details": {
            "activity_kind": "conversion",
            "activity_slug": "ordered_impression_kit",
            "cases": [
              {
                "column": "ordered_impression_kit_at",
                "filter": {
                  "operator": "is_not_null",
                  "value": null,
                },
                "value": 1,
                "value_kind": "number",
              },
            ],
            "kind": "ifttt",
            "value": 0,
            "value_kind": "number",
          },
          "source_kind": "computed",
          "type": "number",
        },
        {
          "filters": [],
          "id": "days_from_session_to_ordered_impression_kit",
          "label": "days_from_session_to_ordered_impression_kit",
          "name": "",
          "output": true,
          "source_details": {
            "activity_kind": "conversion",
            "activity_slug": "ordered_impression_kit",
            "column": "first_session_at",
            "kind": "time_between",
            "second_column": "ordered_impression_kit_at",
            "segmentation": "days",
          },
          "source_kind": "computed",
          "type": "number",
        },
      ],
      "config": {
        "activity_stream": "mv_activity_stream",
        "customer_table": "v_customers",
        "enrichment_column": null,
        "enrichment_table": null,
        "has_source": true,
      },
      "did": true,
      "filters": [],
      "id": "ordered_impression_kit_UUID",
      "kind": "conversion",
      "occurence_after_time": false,
      "occurrence": "first",
      "occurrence_value": 1,
      "relationships": [],
      "slug": "ordered_impression_kit",
    },
  ],
  "_customer_table_columns": [
    {
      "filters": [],
      "id": "customer_name",
      "label": "name",
      "name": "name",
      "output": true,
      "source_details": {
        "table": "v_customers",
      },
      "source_kind": "customer",
      "type": "string",
    },
  ],
  "_limiting_activities": [
    {
      "_columns": [
        {
          "filters": [],
          "id": "email",
          "label": "email",
          "name": "customer",
          "output": true,
          "source_details": {
            "activity_kind": "limiting",
            "activity_slug": "session",
          },
          "source_kind": "activity",
          "type": "string",
        },
        {
          "filters": [],
          "id": "segment_id",
          "label": "segment_id",
          "name": "source_id",
          "output": true,
          "source_details": {
            "activity_id": "session_UUID",
            "activity_kind": "limiting",
          },
          "source_kind": "activity",
          "type": "string",
        },
        {
          "filters": [
            {
              "operator": ">",
              "value": "2018-10-01",
            },
          ],
          "id": "first_session_at",
          "label": "first_session_at",
          "name": "ts",
          "output": true,
          "source_details": {
            "activity_kind": "limiting",
            "activity_slug": "session",
          },
          "source_kind": "activity",
          "type": "timestamp",
        },
        {
          "filters": [],
          "id": "id_device",
          "label": "device",
          "name": "device",
          "output": true,
          "source_details": {
            "activity_kind": "limiting",
            "activity_slug": "session",
          },
          "source_kind": "enrichment",
          "type": "string",
        },
        {
          "filters": [],
          "id": "session_source",
          "label": "session_source",
          "name": "session_source",
          "output": true,
          "source_details": {
            "activity_kind": "limiting",
            "activity_slug": "session",
          },
          "source_kind": "enrichment",
          "type": "string",
        },
        {
          "filters": [],
          "id": "session_kind",
          "label": "session_kind",
          "name": "session_kind",
          "output": true,
          "source_details": {
            "activity_kind": "limiting",
            "activity_slug": "session",
          },
          "source_kind": "enrichment",
          "type": "string",
        },
      ],
      "config": {
        "activity_stream": "mv_activity_stream",
        "customer_table": "v_customers",
        "enrichment_column": "enriched_activity_id",
        "enrichment_table": "mv_enriched_pages",
        "has_source": true,
      },
      "did": true,
      "filters": [],
      "id": "session_UUID",
      "kind": "limiting",
      "occurence_after_time": false,
      "occurrence": "first",
      "occurrence_value": 1,
      "relationships": [],
      "slug": "session",
    },
    {
      "_columns": [
        {
          "filters": [],
          "id": "survey_email_captured_at",
          "label": "survey_email_captured_at",
          "name": "ts",
          "output": true,
          "source_details": {
            "activity_kind": "limiting",
            "activity_slug": "survey_email_captured",
          },
          "source_kind": "activity",
          "type": "timestamp",
        },
      ],
      "config": {
        "activity_stream": "mv_activity_stream",
        "customer_table": "v_customers",
        "enrichment_column": null,
        "enrichment_table": "mv_enrich_uranium",
        "has_source": true,
      },
      "did": true,
      "filters": [],
      "kind": "limiting",
      "occurence_after_time": false,
      "occurrence": "first",
      "occurrence_value": 1,
      "relationships": [],
      "slug": "survey_email_captured",
    },
  ],
  "_staged_activity": {
    "_columns": [],
    "_fieldIndex": 0,
    "_isEdit": false,
    "config": {},
    "did": true,
    "filters": [],
    "kind": "limiting",
    "occurence_after_time": false,
    "occurrence": "all",
    "occurrence_value": 1,
    "relationships": [],
    "slug": null,
  },
  "_staged_computed_column": {
    "_auto_generated_by": null,
    "filters": [],
    "id": null,
    "label": "",
    "name": "",
    "output": true,
    "source_details": {},
    "source_kind": "computed",
    "type": "string",
  },
  "_staged_filters": [
    {
      "kind": "value",
      "operator": "",
      "or_null": false,
      "value": "",
    },
  ],
  "_staged_group_by": {
    "_column_ids": [],
    "columns": [],
    "computed_columns": [],
    "metrics": [
      {
        "_auto_generated_by": null,
        "_pre_pivot_column_id": undefined,
        "_pre_pivot_column_label": undefined,
        "agg_function": "COUNT_ALL",
        "column_id": null,
        "filters": [],
        "id": "metrics_total_events",
        "label": "Count Records",
        "output": true,
        "pivot": [],
        "type": "number",
      },
    ],
    "name": "",
    "order": [],
    "parent_filters": [],
    "pivot": [],
    "slug": null,
  },
  "_staged_group_by_metric": {
    "_auto_generated_by": null,
    "_pre_pivot_column_id": undefined,
    "_pre_pivot_column_label": undefined,
    "agg_function": "COUNT_ALL",
    "column_id": null,
    "filters": [],
    "id": "metrics_total_events",
    "label": "Count Records",
    "output": true,
    "pivot": [],
    "type": "number",
  },
  "_staged_order_by": [],
  "activity_stream": "mv_activity_stream",
  "all_groups": [
    {
      "column_ids": [
        "device",
      ],
      "computed_columns": [],
      "metrics": [
        {
          "agg_function": "COUNT",
          "column_id": "*",
          "filters": [],
          "id": "metrics_total_events",
          "label": "total_events",
          "pivot": [],
        },
        {
          "agg_function": "SUM",
          "column_id": "converted_to_survey_wizard_completed",
          "filters": [],
          "id": "metrics_total_survey_wizard_completed",
          "label": "total_survey_wizard_completed",
          "pivot": [],
        },
      ],
      "name": "Group by Device",
      "order": [
        {
          "column_id": "metrics_total_events",
          "order_direction": "asc",
        },
      ],
      "pivot": [],
      "slug": "group_by_device",
    },
  ],
  "order": [
    {
      "column_id": "decimated_session",
      "order_direction": "desc",
    },
  ],
}
`;
